software/py/blood_graph.py

```diff
@@ -74,77 +298,99 @@ def get_tg_dim(self, traces):
     self.img = Image.new("RGB", (self.img_width, self.img_height), "black")
     self.pixel = self.img.load()
+    return  
   
-  
-  # private function
   # mark the trace on output image
-  def mark_trace(self, trace):
+  def __mark_trace(self, trace):
 
-    # ignore trace outside the time range
-    if trace["timestamp"] < self.start_time or trace["timestamp"] >= self.end_time:
+    # ignore trace outside the cycle range
+    if trace["cycle"] < self.start_cycle or trace["cycle"] >= self.end_cycle:
       return
 
     # determine pixel location
-    cycle = (trace["timestamp"]-self.start_time)//self.timestep
+    cycle = (trace["cycle"] - self.start_cycle)
     col = cycle % self.img_width
     floor = cycle // self.img_width
     tg_x = trace["x"] - self.xmin 
     tg_y = trace["y"] - self.ymin
     row = floor*(2+(self.xdim*self.ydim)) + (tg_x+(tg_y*self.xdim))
 
+
     # determine color
-    if trace["stall"] == "stall_depend":
-      self.pixel[col,row] = (0xdc, 0x14, 0x3c) # crimson
-    elif trace["stall"] == "stall_depend_local_load":
-      self.pixel[col,row] = (0xff, 0x45, 0x00) # orange
-    elif trace["stall"] == "stall_depend_remote_load":
-      self.pixel[col,row] = (0xff, 0x00, 0x00) # red
-    elif trace["stall"] == "stall_depend_local_remote_load":
-      self.pixel[col,row] = (0x80, 0x00, 0x00) # maroon
-    elif trace["stall"] == "stall_fp_remote_load":
-      self.pixel[col,row] = (0x00, 0x80, 0x00) # green
-    elif trace["stall"] == "stall_fp_local_load":
-      self.pixel[col,row] = (0x00, 0x64, 0x00) # darkgreen
-    elif trace["stall"] == "stall_force_wb":
-      self.pixel[col,row] = (0x20, 0xb2, 0xaa) # lightseagreen
-    elif trace["stall"] == "stall_ifetch_wait":
-      self.pixel[col,row] = (0x00, 0x00, 0xff) # blue
-    elif trace["stall"] == "stall_icache_store":
-      self.pixel[col,row] = (0x00, 0x00, 0xff) # grey
-    elif trace["stall"] == "stall_lr_aq":
-      self.pixel[col,row] = (0xff, 0xd7, 0x00) # gold
-    elif trace["stall"] == "stall_md":
-      self.pixel[col,row] = (0x80, 0x00, 0x80) # purple
-    elif trace["stall"] == "stall_remote_req":
-      self.pixel[col,row] = (0x00, 0xff, 0xff) # cyan
-    elif trace["stall"] == "stall_local_flw":
-      self.pixel[col,row] = (0x00, 0xff, 0x00) # lime
-    elif trace["stall"] == "no_stall":
-      self.pixel[col,row] = (0xff, 0xff, 0xff) # white
-      
+    if trace["operation"] in self.stall_bubble_color.keys():
+      self.pixel[col,row] = self.stall_bubble_color[trace["operation"]]
+    elif trace["operation"] in self.instr_list:
+      self.pixel[col,row] = self.unified_instr_color
+    elif trace["operation"] in self.fp_instr_list:
+      self.pixel[col,row] = self.unified_fp_instr_color
+    elif trace["operation"] in self.unknown_list:
+      self.pixel[col,row] = self.unknown_color
+    else:
+      print ("Error: invalid operaiton in operation log: " + trace["operation"])
+      sys.exit()
+    return
```

Comment:
Raise an exception instead of calling sys.exit()

Reply:
Good idea, thanks a lot for pointing it out! Done!